src/ServiceControl.Monitoring.Data.Tests/RawDataReporterTests.cs

```diff
@@ -138,6 +139,59 @@ Task Report(byte[] payload)
             AssertValues(new long[] { 1, 2 });
         }
 
+        [Test]
+        public async Task When_highthrouput_endpoint_shuts_down()
+        {
+            const int testSize = 10000;
+            var queue = new ConcurrentQueue<byte[]>();
+
+            Task Send(byte[] data)
+            {
+                queue.Enqueue(data);
+                return Task.Delay(TimeSpan.FromMilliseconds(100)); // delay to make it realistically long
+            }
+
+            var reporter = new RawDataReporter(Send, buffer, WriteEntriesValues);
+            reporter.Start();
+
+            var writer = Task.Run(() =>
```

Comment:
I think this can be a source of the hanging test. What can happen is that before this task gets started the `reporter.Stop` has already finished. 
What will happen then is that `buffer.TryWrite(i)` will start returning false after the buffer size has been reached. At that point the test is stuck. The writes will never succeed and the `Task.WenAll(writer, stop)` will not continue as well.

Reply:
@tmasternak Wow, this must be it.
@danielmarbach I like the synchronization point approach.